### ISSUE TITLE
initial-delivery-count not required on receiver ATTACH frame

### ISSUE THREAD

**Issue by annatisch** [Azure/azure-uamqp-c#247]
According to the AMQP spec here:
http://docs.oasis-open.org/amqp/core/v1.0/os/amqp-core-transport-v1.0-os.html#type-attach

The initial-delivery-count field is only required for a sender role, and ignored for a receiver.
However the code currently only enforces it for a receiver role:
https://github.com/Azure/azure-uamqp-c/blob/master/src/link.c#L330

I tried simply changing this statement to `if ((link_instance->role == role_sender) &&` however this didn't work because it seems the the ATTACH frames for the $cbs links do not set the initial-delivery-count for the sender - possibly a service bug?

Currently I've modified the code to simply not enforce the presence of initial-delivery-count for either sender or receiver - but I doubt this is the correct solution. I'm happy to PR a fix if you have any thoughts/opinions on the correct approach :)

**Comment by dcristoloveanu** [comment]
Hi @annatisch 

Indeed the AMQP spec states that "This MUST NOT be null if role is sender, and it is ignored if the role is receiver. See subsection 2.6.7"

The [https://github.com/Azure/azure-uamqp-c/blob/master/src/link.c#L330](https://github.com/Azure/azure-uamqp-c/blob/master/src/link.c#L330) is performing the check in the context of an Attach frame being received on that link.

So if the local link endpoint role is receiver (meaning that the peer is a sender) then we do expect to have an initial delivery count on the ATTACH.

Are you seeing cases where the sender is sending an ATTACH with no initial-delivery-count?
I think that is an AMQP spec violation. If that is the case, which service is it so we can follow up properly with the owners.

Cheers,
/Dan

**Comment by dcristoloveanu** [comment]
@annatisch Indeed, if we (client) initiated a receiver link and we receive an attach frame without initial-delivery-count that is essentially violating the spec.

"initial-delivery-count the sender’s initial value for delivery-count
This MUST NOT be null if role is sender, and it is ignored if the role is receiver. See subsection
2.6.7."

But I believe 2.6.3 is also stating
"If there is no pre-existing terminus, and the peer does not wish to create a new one, this is indicated by setting the
local terminus (source or target as appropriate) to null."

Which means that well, the server should send an ATTACH with NULL source/target in this case, which it does.
And it would be awkward for the broker (service) to send some fake initial delivery count. SO I think the service is actually right on not sending an initial delivery count.

Overall, after re-reviewing this, I think uAMQP is too aggressive here with regard to that part of the spec which says that initial-delivery-count must be set.

So I will change uAMQP to allow that to be NULL when there is a NULL source/target indicating that the peer does not wish to create the link endpoint.

HTH,
/Dan

**Comment by dcristoloveanu** [comment]
Hi @annatisch,

I was attempting to fix this, but while doing so I realized I might have missed something :-).

So I will just state what I realized first:
- The check for the initial-delivery-count does make sense, since under normal circumstances (when the broker does create a link endpoint) the initial-delivery-count must be there on the ATTACH frames sent by the sender.
uAMQP does have the check
- For the cases that the broker does not wish to create the link endpoint (because a redirect is intended), the link should anyhow not be brought up on the client side (so the checks for initial-delivery-count etc. still make sense)
uAMQP behaves this way.
- For the redirect cases the link endpoint needs to be remembered for the detach information and this does happen at the session level where we have a list of all link endpoints.

If you could help me with the steps on repro-ing this it would be awesome!
I tried having a test to repro where the broker sends:
- the ATTACH frame with NULL initial-delivery-count and the redirect information was
- the DETACH with the redirect info
This seemed to work fine, as in the test did pick up the redirect information.

So it must be that you're seeing a different type of failure! I really would like to fix this so ... "Help in getting the repro :-)".

Cheers,
/Dan

**Comment by annatisch** [comment]
Thanks for looking into this @dcristoloveanu!
I'm trying to remember under which circumstances I saw this behavior.... I think it was attempting to open a receiver against an eventhub that already had an epoch receiver running.

I will run through my tests again to try and reproduce :)

**Comment by dcristoloveanu** [comment]
@annatisch Sorry for the slight delay.
So just for my fully understanding this:
Link 7 in the log above gets detached by the service side because of the limit you mentioned.
On the client side the fact that the link state changed gets propagated to the message receiver, which in turn changes its state to be ERROR and notifies its user.

Up to now seems OK. The user of the receiver needs to decide what to do at this point. There are probably several things that the client can do:
1. Close the connection (and potentially reopen it). This is obviously not desired as the other links are all good.
2. Close the receiver and reopen it. Although better than 1., still does not make sense as the link would be most likely detached again because of the limit.
3. Look at the error and decide to inform the user that the receiving cannot happen. Now I can understand how this is hard to do given that the detach handler will be called later. I can reverse the order of the state update and link_detach handler to make things easier :-).

Cheers,
/Dan

**Comment by dcristoloveanu** [comment]
I put the change in the link_detach_before_state_change branch. Let me know if it helps then I can merge it in.

Cheers,
/Dan

**Comment by annatisch** [comment]
Thanks! I will try to test today :)
